zk_toolbox/crates/zk_inception/src/utils/ports.rs

```diff
@@ -0,0 +1,303 @@
+use std::{
+    collections::{HashMap, HashSet},
+    fmt, fs,
+    ops::Range,
+    path::Path,
+};
+
+use anyhow::Result;
+use common::logger;
+use config::EcosystemConfig;
+use serde_yaml::Value;
+use xshell::Shell;
+
+/// Ecosystem-level directories to include in the scan
+const INCLUDE_DIRS: &[&str] = &["chains", "configs"];
+/// Skip directories with these names at all levels
+const SKIP_DIRS: &[&str] = &["db", "volumes"];
+
+pub struct EcosystemPorts {
+    pub ports: HashMap<u16, Vec<String>>,
+}
+
+impl EcosystemPorts {
+    pub fn new() -> Self {
+        Self {
+            ports: HashMap::from([
+                (3000, vec!["Observability".to_string()]),
+                (3052, vec!["External node gateway".to_string()]),
+            ]),
+        }
+    }
+
+    pub fn get_assigned_ports(&self) -> HashSet<u16> {
+        self.ports.keys().cloned().collect()
+    }
+
+    pub fn is_port_assigned(&self, port: u16) -> bool {
+        self.ports.contains_key(&port)
+    }
+
+    pub fn add_port_info(&mut self, port: u16, info: String) {
+        self.ports.entry(port).or_default().push(info);
+    }
+
+    pub fn allocate_port(&mut self, range: Range<u16>) -> Result<u16> {
+        for port in range {
+            if !self.is_port_assigned(port) {
+                self.add_port_info(port, "Dynamically allocated".to_string());
+                return Ok(port);
+            }
+        }
+        anyhow::bail!("No available ports in the given range")
+    }
+
+    /// Allocates a set of ports based on given base ports, incrementing by offset if needed.
+    /// Tries base ports first, then base + offset, base + 2*offset, etc., until finding
+    /// available ports or exceeding the range. Returns allocated ports or an error if
+    /// no suitable ports are found.
+    pub fn allocate_ports(
+        &mut self,
+        base_ports: &[u16],
+        range: Range<u16>,
+        offset: u16,
+    ) -> Result<Vec<u16>> {
+        let mut i = 0;
+        loop {
+            let candidate_ports: Vec<u16> =
+                base_ports.iter().map(|&port| port + i * offset).collect();
+
+            // Check if all candidate ports are within the range
+            if candidate_ports.iter().any(|&port| !range.contains(&port)) {
+                anyhow::bail!(
+                    "No suitable ports found within the given range {:?}. Tried {} iterations with offset {}",
+                    range,
+                    i,
+                    offset
+                )
+            }
+
+            // Check if all candidate ports are available
+            if candidate_ports
+                .iter()
+                .all(|&port| !self.is_port_assigned(port))
+            {
+                // Allocate all ports
+                for &port in &candidate_ports {
+                    self.add_port_info(port, "Dynamically allocated".to_string());
```

Comment:
Should we specify for what it was allocated?

Reply:
The way it currently works is that `allocate_port(s)` is used (right after initial ecosystem directory scan) to allocate ports for new services (currently only for explorer ones - worker, api, data_fetcher).
All that is required is to remember newly allocated ports are now busy; there is little chance we will use port descriptions within the same `EcosystemPortsScanner` lifecycle.
Then these ports are saved into configs and next time `EcosystemPortsScanner` is called - it will return ports with the right description.
If we want to keep meaningful descriptions at all times, I'll have to modify allocate_port(s) to include a vector of string description(s). Let me know, I can do it, but it would make usage a bit more complicated, and I don't see a big need for that.